{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kipoi python API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are three basic building blocks in kipoi:\n",
    "\n",
    "- **Source** - provides Models and DataLoaders.\n",
    "- **Model** - makes the prediction given the numpy arrays. \n",
    "- **Dataloader** - loads the data from raw files and transforms them into a form that is directly consumable by the Model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![img](../docs/theme_dir/img/kipoi-workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List of main commands\n",
    "\n",
    "\n",
    "- `kipoi.list_sources()`\n",
    "- `kipoi.get_source()`\n",
    "\n",
    "\n",
    "- `kipoi.list_models()`\n",
    "- `kipoi.list_dataloaders()`\n",
    "\n",
    "\n",
    "- `kipoi.get_model()`\n",
    "- `kipoi.get_dataloader_factory()`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Source\n",
    "\n",
    "Available sources are specified in the config file located at: `~/.kipoi/config.yaml`. Here is an example config file:\n",
    "\n",
    "```yaml\n",
    "model_sources:\n",
    "    kipoi: # default\n",
    "        type: git-lfs # git repository with large file storage (git-lfs)\n",
    "        remote_url: git@github.com:kipoi/models.git # git remote\n",
    "        local_path: ~/.kipoi/models/ # local storage path\n",
    "    gl:\n",
    "        type: git-lfs  # custom model\n",
    "        remote_url: https://i12g-gagneurweb.informatik.tu-muenchen.de/gitlab/gagneurlab/model-zoo.git\n",
    "        local_path: /s/project/model-zoo\n",
    "```\n",
    "\n",
    "There are three different model sources possible: \n",
    "\n",
    "- **`git-lfs`** - git repository with source files tracked normally by git and all the binary files like model weights (located in `files*` directories) are tracked by [git-lfs](https://git-lfs.github.com). \n",
    "  - Requires `git-lfs` to be installed.\n",
    "- **`git`** - all the files including weights (not recommended)\n",
    "- **`local`** - local directory containing models defined in subdirectories\n",
    "\n",
    "For **`git-lfs`** source type, larger files tracked by `git-lfs` will be downloaded into the specified directory `local_path` only after the model has been requested (when invoking `kipoi.get_model()`).\n",
    "\n",
    "#### Note\n",
    "\n",
    "A particular model/dataloader is defined by its source (say `kipoi` or `my_git_models`) and the relative path of the desired model directory from the model source root (say `rbp/`).\n",
    "\n",
    "A directory is considered a model if it contains a `model.yaml` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import kipoi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import logging\n",
    "logging.disable(1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>local_size</th>\n",
       "      <th>n_models</th>\n",
       "      <th>n_dataloaders</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>git-lfs</td>\n",
       "      <td>/homes/rkreuzhu/.kipoi/models/</td>\n",
       "      <td>5.0G</td>\n",
       "      <td>780</td>\n",
       "      <td>780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source     type                        location local_size  n_models  \\\n",
       "0  kipoi  git-lfs  /homes/rkreuzhu/.kipoi/models/       5.0G       780   \n",
       "\n",
       "   n_dataloaders  \n",
       "0            780  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kipoi.list_sources()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = kipoi.get_source(\"kipoi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GitLFSSource(remote_url='git@github.com:kipoi/models.git', local_path='/homes/rkreuzhu/.kipoi/models/')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>model</th>\n",
       "      <th>version</th>\n",
       "      <th>authors</th>\n",
       "      <th>contributors</th>\n",
       "      <th>doc</th>\n",
       "      <th>type</th>\n",
       "      <th>inputs</th>\n",
       "      <th>targets</th>\n",
       "      <th>postproc_score_variants</th>\n",
       "      <th>license</th>\n",
       "      <th>cite_as</th>\n",
       "      <th>trained_on</th>\n",
       "      <th>training_procedure</th>\n",
       "      <th>tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/A549_ENCSR000DDI</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/BE2C_ENCSR000DEB</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/BJ_ENCSR000DEA</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/CMK_ENCSR000DGJ</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/Caco_2_ENCSR000DDO</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM06990_ENCSR000DDN</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM12878_ENCSR000DEY</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM12878_ENCSR000DFT</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM12891_ENCSR000DFO</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM12892_ENCSR000DFN</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM19239_ENCSR000DGH</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/GM19240_ENCSR000DGG</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HCT116_ENCSR000DFS</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HEK293_ENCSR087IEW</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HEK293_ENCSR794HFF</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HGPS_cell_ENCSR000DFA</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HL_60_ENCSR000DDM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HTR_8_SVneo_ENCSR000DDP</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HeLa_S3_ENCSR000DFL</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HepG2_ENCSR530RCY</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/HepG2_ENCSR592WWM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/IMR_90_ENCSR000DEX</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/Ishikawa_ENCSR000DDZ</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/Jurkat_ENCSR000DFP</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/K562_ENCSR000DFK</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/LNCaP_clone_FGC_ENCSR000DFH</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/LNCaP_clone_FGC_ENCSR439FXM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/LNCaP_clone_FGC_ENCSR859PDD</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/MCF_10A_ENCSR000DEV</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>CpGenie/MCF_10A_ENCSR000DEW</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Haoyang Zeng', github='haoyangz'...</td>\n",
       "      <td>[Author(name='Roman Kreuzhuber', github='krrom...</td>\n",
       "      <td>Abstract: DNA methylation plays a crucial role...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>methylation_prob</td>\n",
       "      <td>True</td>\n",
       "      <td>Apache License v2</td>\n",
       "      <td>https://doi.org/10.1093/nar/gkx177</td>\n",
       "      <td>RRBS (restricted representation bisulfite sequ...</td>\n",
       "      <td>RMSprop</td>\n",
       "      <td>[DNA methylation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZEP1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZEP2</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZFHX3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZFX</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZIC1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZIC2</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZIC3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZIC4</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZKSC1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZKSC3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN143</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN148</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN219</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN232</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN282</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN333</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN350</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN384</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN410</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN423</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN524</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN589</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN639</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN652</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN713</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN740</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZN784</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZSC16</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>tf-binding/human/pwm/HOCOMOCO/ZSCA4</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Ivan V. Kulakovskiy', github=Non...</td>\n",
       "      <td>[Author(name='Ziga Avsec', github='avsecz', em...</td>\n",
       "      <td>'''Simple PWM-scanning model\\nPWM database: HO...</td>\n",
       "      <td>keras</td>\n",
       "      <td>seq</td>\n",
       "      <td>pwm_match</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT</td>\n",
       "      <td>https://doi:10.1093/nar/gkv1249</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA binding]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>kipoi</td>\n",
       "      <td>BassetGM12878_Demo</td>\n",
       "      <td>0.1</td>\n",
       "      <td>[Author(name='Nancy Xu', github='xnancy', emai...</td>\n",
       "      <td>[Author(name='Nancy Xu', github='xnancy', emai...</td>\n",
       "      <td>test</td>\n",
       "      <td>keras</td>\n",
       "      <td>[data/genome_data_dir]</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>MIT</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[DNA accessibility]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>780 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    source                                model version  \\\n",
       "0    kipoi             CpGenie/A549_ENCSR000DDI     0.1   \n",
       "1    kipoi             CpGenie/BE2C_ENCSR000DEB     0.1   \n",
       "2    kipoi               CpGenie/BJ_ENCSR000DEA     0.1   \n",
       "3    kipoi              CpGenie/CMK_ENCSR000DGJ     0.1   \n",
       "4    kipoi           CpGenie/Caco_2_ENCSR000DDO     0.1   \n",
       "5    kipoi          CpGenie/GM06990_ENCSR000DDN     0.1   \n",
       "6    kipoi          CpGenie/GM12878_ENCSR000DEY     0.1   \n",
       "7    kipoi          CpGenie/GM12878_ENCSR000DFT     0.1   \n",
       "8    kipoi          CpGenie/GM12891_ENCSR000DFO     0.1   \n",
       "9    kipoi          CpGenie/GM12892_ENCSR000DFN     0.1   \n",
       "10   kipoi          CpGenie/GM19239_ENCSR000DGH     0.1   \n",
       "11   kipoi          CpGenie/GM19240_ENCSR000DGG     0.1   \n",
       "12   kipoi           CpGenie/HCT116_ENCSR000DFS     0.1   \n",
       "13   kipoi           CpGenie/HEK293_ENCSR087IEW     0.1   \n",
       "14   kipoi           CpGenie/HEK293_ENCSR794HFF     0.1   \n",
       "15   kipoi        CpGenie/HGPS_cell_ENCSR000DFA     0.1   \n",
       "16   kipoi            CpGenie/HL_60_ENCSR000DDM     0.1   \n",
       "17   kipoi      CpGenie/HTR_8_SVneo_ENCSR000DDP     0.1   \n",
       "18   kipoi          CpGenie/HeLa_S3_ENCSR000DFL     0.1   \n",
       "19   kipoi            CpGenie/HepG2_ENCSR530RCY     0.1   \n",
       "20   kipoi            CpGenie/HepG2_ENCSR592WWM     0.1   \n",
       "21   kipoi           CpGenie/IMR_90_ENCSR000DEX     0.1   \n",
       "22   kipoi         CpGenie/Ishikawa_ENCSR000DDZ     0.1   \n",
       "23   kipoi           CpGenie/Jurkat_ENCSR000DFP     0.1   \n",
       "24   kipoi             CpGenie/K562_ENCSR000DFK     0.1   \n",
       "25   kipoi  CpGenie/LNCaP_clone_FGC_ENCSR000DFH     0.1   \n",
       "26   kipoi  CpGenie/LNCaP_clone_FGC_ENCSR439FXM     0.1   \n",
       "27   kipoi  CpGenie/LNCaP_clone_FGC_ENCSR859PDD     0.1   \n",
       "28   kipoi          CpGenie/MCF_10A_ENCSR000DEV     0.1   \n",
       "29   kipoi          CpGenie/MCF_10A_ENCSR000DEW     0.1   \n",
       "..     ...                                  ...     ...   \n",
       "753  kipoi   tf-binding/human/pwm/HOCOMOCO/ZEP1     0.1   \n",
       "754  kipoi   tf-binding/human/pwm/HOCOMOCO/ZEP2     0.1   \n",
       "755  kipoi  tf-binding/human/pwm/HOCOMOCO/ZFHX3     0.1   \n",
       "756  kipoi    tf-binding/human/pwm/HOCOMOCO/ZFX     0.1   \n",
       "757  kipoi   tf-binding/human/pwm/HOCOMOCO/ZIC1     0.1   \n",
       "758  kipoi   tf-binding/human/pwm/HOCOMOCO/ZIC2     0.1   \n",
       "759  kipoi   tf-binding/human/pwm/HOCOMOCO/ZIC3     0.1   \n",
       "760  kipoi   tf-binding/human/pwm/HOCOMOCO/ZIC4     0.1   \n",
       "761  kipoi  tf-binding/human/pwm/HOCOMOCO/ZKSC1     0.1   \n",
       "762  kipoi  tf-binding/human/pwm/HOCOMOCO/ZKSC3     0.1   \n",
       "763  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN143     0.1   \n",
       "764  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN148     0.1   \n",
       "765  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN219     0.1   \n",
       "766  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN232     0.1   \n",
       "767  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN282     0.1   \n",
       "768  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN333     0.1   \n",
       "769  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN350     0.1   \n",
       "770  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN384     0.1   \n",
       "771  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN410     0.1   \n",
       "772  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN423     0.1   \n",
       "773  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN524     0.1   \n",
       "774  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN589     0.1   \n",
       "775  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN639     0.1   \n",
       "776  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN652     0.1   \n",
       "777  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN713     0.1   \n",
       "778  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN740     0.1   \n",
       "779  kipoi  tf-binding/human/pwm/HOCOMOCO/ZN784     0.1   \n",
       "780  kipoi  tf-binding/human/pwm/HOCOMOCO/ZSC16     0.1   \n",
       "781  kipoi  tf-binding/human/pwm/HOCOMOCO/ZSCA4     0.1   \n",
       "783  kipoi                   BassetGM12878_Demo     0.1   \n",
       "\n",
       "                                               authors  \\\n",
       "0    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "1    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "2    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "3    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "4    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "5    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "6    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "7    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "8    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "9    [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "10   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "11   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "12   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "13   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "14   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "15   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "16   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "17   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "18   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "19   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "20   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "21   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "22   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "23   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "24   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "25   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "26   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "27   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "28   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "29   [Author(name='Haoyang Zeng', github='haoyangz'...   \n",
       "..                                                 ...   \n",
       "753  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "754  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "755  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "756  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "757  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "758  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "759  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "760  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "761  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "762  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "763  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "764  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "765  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "766  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "767  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "768  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "769  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "770  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "771  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "772  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "773  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "774  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "775  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "776  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "777  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "778  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "779  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "780  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "781  [Author(name='Ivan V. Kulakovskiy', github=Non...   \n",
       "783  [Author(name='Nancy Xu', github='xnancy', emai...   \n",
       "\n",
       "                                          contributors  \\\n",
       "0    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "1    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "2    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "3    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "4    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "5    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "6    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "7    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "8    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "9    [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "10   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "11   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "12   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "13   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "14   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "15   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "16   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "17   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "18   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "19   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "20   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "21   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "22   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "23   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "24   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "25   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "26   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "27   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "28   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "29   [Author(name='Roman Kreuzhuber', github='krrom...   \n",
       "..                                                 ...   \n",
       "753  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "754  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "755  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "756  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "757  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "758  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "759  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "760  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "761  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "762  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "763  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "764  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "765  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "766  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "767  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "768  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "769  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "770  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "771  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "772  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "773  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "774  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "775  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "776  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "777  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "778  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "779  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "780  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "781  [Author(name='Ziga Avsec', github='avsecz', em...   \n",
       "783  [Author(name='Nancy Xu', github='xnancy', emai...   \n",
       "\n",
       "                                                   doc   type  \\\n",
       "0    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "1    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "2    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "3    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "4    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "5    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "6    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "7    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "8    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "9    Abstract: DNA methylation plays a crucial role...  keras   \n",
       "10   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "11   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "12   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "13   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "14   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "15   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "16   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "17   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "18   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "19   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "20   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "21   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "22   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "23   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "24   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "25   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "26   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "27   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "28   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "29   Abstract: DNA methylation plays a crucial role...  keras   \n",
       "..                                                 ...    ...   \n",
       "753  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "754  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "755  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "756  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "757  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "758  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "759  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "760  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "761  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "762  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "763  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "764  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "765  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "766  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "767  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "768  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "769  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "770  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "771  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "772  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "773  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "774  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "775  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "776  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "777  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "778  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "779  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "780  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "781  '''Simple PWM-scanning model\\nPWM database: HO...  keras   \n",
       "783                                               test  keras   \n",
       "\n",
       "                     inputs           targets  postproc_score_variants  \\\n",
       "0                       seq  methylation_prob                     True   \n",
       "1                       seq  methylation_prob                     True   \n",
       "2                       seq  methylation_prob                     True   \n",
       "3                       seq  methylation_prob                     True   \n",
       "4                       seq  methylation_prob                     True   \n",
       "5                       seq  methylation_prob                     True   \n",
       "6                       seq  methylation_prob                     True   \n",
       "7                       seq  methylation_prob                     True   \n",
       "8                       seq  methylation_prob                     True   \n",
       "9                       seq  methylation_prob                     True   \n",
       "10                      seq  methylation_prob                     True   \n",
       "11                      seq  methylation_prob                     True   \n",
       "12                      seq  methylation_prob                     True   \n",
       "13                      seq  methylation_prob                     True   \n",
       "14                      seq  methylation_prob                     True   \n",
       "15                      seq  methylation_prob                     True   \n",
       "16                      seq  methylation_prob                     True   \n",
       "17                      seq  methylation_prob                     True   \n",
       "18                      seq  methylation_prob                     True   \n",
       "19                      seq  methylation_prob                     True   \n",
       "20                      seq  methylation_prob                     True   \n",
       "21                      seq  methylation_prob                     True   \n",
       "22                      seq  methylation_prob                     True   \n",
       "23                      seq  methylation_prob                     True   \n",
       "24                      seq  methylation_prob                     True   \n",
       "25                      seq  methylation_prob                     True   \n",
       "26                      seq  methylation_prob                     True   \n",
       "27                      seq  methylation_prob                     True   \n",
       "28                      seq  methylation_prob                     True   \n",
       "29                      seq  methylation_prob                     True   \n",
       "..                      ...               ...                      ...   \n",
       "753                     seq         pwm_match                     True   \n",
       "754                     seq         pwm_match                     True   \n",
       "755                     seq         pwm_match                     True   \n",
       "756                     seq         pwm_match                     True   \n",
       "757                     seq         pwm_match                     True   \n",
       "758                     seq         pwm_match                     True   \n",
       "759                     seq         pwm_match                     True   \n",
       "760                     seq         pwm_match                     True   \n",
       "761                     seq         pwm_match                     True   \n",
       "762                     seq         pwm_match                     True   \n",
       "763                     seq         pwm_match                     True   \n",
       "764                     seq         pwm_match                     True   \n",
       "765                     seq         pwm_match                     True   \n",
       "766                     seq         pwm_match                     True   \n",
       "767                     seq         pwm_match                     True   \n",
       "768                     seq         pwm_match                     True   \n",
       "769                     seq         pwm_match                     True   \n",
       "770                     seq         pwm_match                     True   \n",
       "771                     seq         pwm_match                     True   \n",
       "772                     seq         pwm_match                     True   \n",
       "773                     seq         pwm_match                     True   \n",
       "774                     seq         pwm_match                     True   \n",
       "775                     seq         pwm_match                     True   \n",
       "776                     seq         pwm_match                     True   \n",
       "777                     seq         pwm_match                     True   \n",
       "778                     seq         pwm_match                     True   \n",
       "779                     seq         pwm_match                     True   \n",
       "780                     seq         pwm_match                     True   \n",
       "781                     seq         pwm_match                     True   \n",
       "783  [data/genome_data_dir]              None                    False   \n",
       "\n",
       "               license                             cite_as  \\\n",
       "0    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "1    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "2    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "3    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "4    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "5    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "6    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "7    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "8    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "9    Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "10   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "11   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "12   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "13   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "14   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "15   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "16   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "17   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "18   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "19   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "20   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "21   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "22   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "23   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "24   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "25   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "26   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "27   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "28   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "29   Apache License v2  https://doi.org/10.1093/nar/gkx177   \n",
       "..                 ...                                 ...   \n",
       "753                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "754                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "755                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "756                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "757                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "758                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "759                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "760                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "761                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "762                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "763                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "764                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "765                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "766                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "767                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "768                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "769                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "770                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "771                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "772                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "773                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "774                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "775                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "776                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "777                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "778                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "779                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "780                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "781                MIT     https://doi:10.1093/nar/gkv1249   \n",
       "783                MIT                                None   \n",
       "\n",
       "                                            trained_on training_procedure  \\\n",
       "0    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "1    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "2    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "3    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "4    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "5    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "6    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "7    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "8    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "9    RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "10   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "11   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "12   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "13   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "14   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "15   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "16   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "17   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "18   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "19   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "20   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "21   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "22   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "23   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "24   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "25   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "26   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "27   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "28   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "29   RRBS (restricted representation bisulfite sequ...            RMSprop   \n",
       "..                                                 ...                ...   \n",
       "753                                               None               None   \n",
       "754                                               None               None   \n",
       "755                                               None               None   \n",
       "756                                               None               None   \n",
       "757                                               None               None   \n",
       "758                                               None               None   \n",
       "759                                               None               None   \n",
       "760                                               None               None   \n",
       "761                                               None               None   \n",
       "762                                               None               None   \n",
       "763                                               None               None   \n",
       "764                                               None               None   \n",
       "765                                               None               None   \n",
       "766                                               None               None   \n",
       "767                                               None               None   \n",
       "768                                               None               None   \n",
       "769                                               None               None   \n",
       "770                                               None               None   \n",
       "771                                               None               None   \n",
       "772                                               None               None   \n",
       "773                                               None               None   \n",
       "774                                               None               None   \n",
       "775                                               None               None   \n",
       "776                                               None               None   \n",
       "777                                               None               None   \n",
       "778                                               None               None   \n",
       "779                                               None               None   \n",
       "780                                               None               None   \n",
       "781                                               None               None   \n",
       "783                                               None               None   \n",
       "\n",
       "                    tags  \n",
       "0      [DNA methylation]  \n",
       "1      [DNA methylation]  \n",
       "2      [DNA methylation]  \n",
       "3      [DNA methylation]  \n",
       "4      [DNA methylation]  \n",
       "5      [DNA methylation]  \n",
       "6      [DNA methylation]  \n",
       "7      [DNA methylation]  \n",
       "8      [DNA methylation]  \n",
       "9      [DNA methylation]  \n",
       "10     [DNA methylation]  \n",
       "11     [DNA methylation]  \n",
       "12     [DNA methylation]  \n",
       "13     [DNA methylation]  \n",
       "14     [DNA methylation]  \n",
       "15     [DNA methylation]  \n",
       "16     [DNA methylation]  \n",
       "17     [DNA methylation]  \n",
       "18     [DNA methylation]  \n",
       "19     [DNA methylation]  \n",
       "20     [DNA methylation]  \n",
       "21     [DNA methylation]  \n",
       "22     [DNA methylation]  \n",
       "23     [DNA methylation]  \n",
       "24     [DNA methylation]  \n",
       "25     [DNA methylation]  \n",
       "26     [DNA methylation]  \n",
       "27     [DNA methylation]  \n",
       "28     [DNA methylation]  \n",
       "29     [DNA methylation]  \n",
       "..                   ...  \n",
       "753        [DNA binding]  \n",
       "754        [DNA binding]  \n",
       "755        [DNA binding]  \n",
       "756        [DNA binding]  \n",
       "757        [DNA binding]  \n",
       "758        [DNA binding]  \n",
       "759        [DNA binding]  \n",
       "760        [DNA binding]  \n",
       "761        [DNA binding]  \n",
       "762        [DNA binding]  \n",
       "763        [DNA binding]  \n",
       "764        [DNA binding]  \n",
       "765        [DNA binding]  \n",
       "766        [DNA binding]  \n",
       "767        [DNA binding]  \n",
       "768        [DNA binding]  \n",
       "769        [DNA binding]  \n",
       "770        [DNA binding]  \n",
       "771        [DNA binding]  \n",
       "772        [DNA binding]  \n",
       "773        [DNA binding]  \n",
       "774        [DNA binding]  \n",
       "775        [DNA binding]  \n",
       "776        [DNA binding]  \n",
       "777        [DNA binding]  \n",
       "778        [DNA binding]  \n",
       "779        [DNA binding]  \n",
       "780        [DNA binding]  \n",
       "781        [DNA binding]  \n",
       "783  [DNA accessibility]  \n",
       "\n",
       "[780 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kipoi.list_models()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model\n",
    "\n",
    "Let's choose to use the `rbp_eclip/UPF1` model from kipoi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "# Note. Install all the dependencies for that model using\n",
    "# kipoi env install \n",
    "model = kipoi.get_model(\"rbp_eclip/UPF1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Available fields:\n",
    "\n",
    "#### Model\n",
    "\n",
    "- type\n",
    "- args\n",
    "- info\n",
    "  - authors\n",
    "  - name\n",
    "  - version\n",
    "  - tags\n",
    "  - doc\n",
    "- schema\n",
    "  - inputs\n",
    "  - targets\n",
    "- default_dataloader - loaded dataloader class\n",
    "\n",
    "\n",
    "- predict_on_batch()\n",
    "- source\n",
    "- source_dir\n",
    "- pipeline\n",
    "  - predict()\n",
    "  - predict_example()\n",
    "  - predict_generator()\n",
    "  \n",
    "#### Dataloader\n",
    "\n",
    "- type\n",
    "- defined_as\n",
    "- args\n",
    "- info (same as for the model)\n",
    "- output_schema\n",
    "  - inputs\n",
    "  - targets\n",
    "  - metadata\n",
    "\n",
    "\n",
    "- source\n",
    "- source_dir\n",
    "- example_kwargs\n",
    "- init_example()\n",
    "- batch_iter()\n",
    "- batch_train_iter()\n",
    "- batch_predict_iter()\n",
    "- load_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<kipoi.model.KerasModel at 0x2ada1ede6cc0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'keras'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ModelInfo(authors=[Author(name='Ziga Avsec', github='avsecz', email=None)], doc='\\'RBP binding model from Avsec et al: \"Modeling positional effects of regulatory sequences with spline transformations increases prediction accuracy of deep neural networks\". \\'\\n', name=None, version='0.1', license='MIT', tags=['RNA binding'], contributors=[Author(name='Ziga Avsec', github='avsecz', email=None)], cite_as='https://doi.org/10.1093/bioinformatics/btx727', trained_on='RBP occupancy peaks measured by eCLIP-seq (Van Nostrand et al., 2016 - https://doi.org/10.1038/nmeth.3810), https://github.com/gagneurlab/Manuscript_Avsec_Bioinformatics_2017\\n', training_procedure='Single task training with ADAM')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.1'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.info.version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dist_exon_intron': ArraySchema(shape=(1, 10), doc='Distance the nearest exon_intron (splice donor) site transformed with B-splines', name='dist_exon_intron', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_gene_end': ArraySchema(shape=(1, 10), doc='Distance the nearest gene end transformed with B-splines', name='dist_gene_end', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_gene_start': ArraySchema(shape=(1, 10), doc='Distance the nearest gene start transformed with B-splines', name='dist_gene_start', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_intron_exon': ArraySchema(shape=(1, 10), doc='Distance the nearest intron_exon (splice acceptor) site transformed with B-splines', name='dist_intron_exon', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_polya': ArraySchema(shape=(1, 10), doc='Distance the nearest Poly-A site transformed with B-splines', name='dist_polya', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_start_codon': ArraySchema(shape=(1, 10), doc='Distance the nearest start codon transformed with B-splines', name='dist_start_codon', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_stop_codon': ArraySchema(shape=(1, 10), doc='Distance the nearest stop codon transformed with B-splines', name='dist_stop_codon', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'dist_tss': ArraySchema(shape=(1, 10), doc='Distance the nearest TSS site transformed with B-splines', name='dist_tss', special_type=None, associated_metadata=[], column_labels=None),\n",
       " 'seq': ArraySchema(shape=(101, 4), doc='One-hot encoded RNA sequence', name='seq', special_type=<ArraySpecialType.DNASeq: 'DNASeq'>, associated_metadata=[], column_labels=None)}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(model.schema.inputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ArraySchema(shape=(1,), doc='Predicted binding strength', name=None, special_type=None, associated_metadata=[], column_labels=None)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.schema.targets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Default dataloader\n",
    "\n",
    "Model already has the default dataloder present. To use it, specify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/homes/rkreuzhu/.kipoi/models/rbp_eclip/UPF1'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.source_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dataloader.SeqDistDataset"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.default_dataloader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Info(authors=[Author(name='Ziga Avsec', github='avsecz', email=None)], doc='RBP binding model taking as input 101nt long sequence as well as 8 distances to nearest genomic landmarks -  tss, poly-A, exon-intron boundary, intron-exon boundary, start codon, stop codon, gene start, gene end\\n', name=None, version='0.1', license='MIT', tags=[])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.default_dataloader.info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predict_on_batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method KerasModel.predict_on_batch of <kipoi.model.KerasModel object at 0x2ada1ede6cc0>>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict_on_batch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pipeline\n",
    "\n",
    "Pipeline object will take the dataloader arguments and run the whole pipeline:\n",
    "\n",
    "```\n",
    "dataloader arguments --Dataloader-->  numpy arrays --Model--> prediction\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#model.pipeline.predict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#model.pipeline.predict_generator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GitLFSSource(remote_url='git@github.com:kipoi/models.git', local_path='/homes/rkreuzhu/.kipoi/models/')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Model source\n",
    "model.source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/homes/rkreuzhu/.kipoi/models/rbp_eclip/UPF1'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# model location directory\n",
    "model.source_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DataLoader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "DataLoader = kipoi.get_dataloader_factory(\"rbp_eclip/UPF1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A dataloader will most likely require input arguments in which the input files are defined, for example input fasta files or bed files, based on which the model input is generated. There are several options where the dataloader input keyword arguments are displayed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Args:\n",
      "        intervals_file: file path; tsv file\n",
      "            Assumes bed-like `chrom start end id score strand` format.\n",
      "        fasta_file: file path; Genome sequence\n",
      "        gtf_file: file path; Genome annotation GTF file.\n",
      "        filter_protein_coding: Considering genomic landmarks only for protein coding genes\n",
      "        preproc_transformer: file path; tranformer used for pre-processing.\n",
      "        target_file: file path; path to the targets\n",
      "        batch_size: int\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# Display information about the dataloader\n",
    "print(DataLoader.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keyword argument: `intervals_file`\n",
      "    doc: bed6 file with `chrom start end id score strand` columns\n",
      "    type: str\n",
      "    optional: False\n",
      "    example: example_files/intervals.bed\n",
      "Keyword argument: `fasta_file`\n",
      "    doc: Reference genome sequence\n",
      "    type: str\n",
      "    optional: False\n",
      "    example: example_files/hg38_chr22.fa\n",
      "Keyword argument: `gtf_file`\n",
      "    doc: file path; Genome annotation GTF file\n",
      "    type: str\n",
      "    optional: False\n",
      "    example: example_files/gencode.v24.annotation_chr22.gtf\n",
      "Keyword argument: `filter_protein_coding`\n",
      "    doc: Considering genomic landmarks only for protein coding genes when computing the distances to the nearest genomic landmark.\n",
      "    type: str\n",
      "    optional: True\n",
      "    example: True\n",
      "Keyword argument: `target_file`\n",
      "    doc: path to the targets (txt) file\n",
      "    type: str\n",
      "    optional: True\n",
      "    example: example_files/targets.tsv\n",
      "Keyword argument: `use_linecache`\n",
      "    doc: if True, use linecache https://docs.python.org/3/library/linecache.html to access bed file rows\n",
      "    type: str\n",
      "    optional: True\n",
      "--------------------------------------------------------------------------------\n",
      "Example keyword arguments are: {'gtf_file': 'example_files/gencode.v24.annotation_chr22.gtf', 'fasta_file': 'example_files/hg38_chr22.fa', 'target_file': 'example_files/targets.tsv', 'intervals_file': 'example_files/intervals.bed', 'filter_protein_coding': True}\n"
     ]
    }
   ],
   "source": [
    "# Alternatively the dataloader keyword arguments can be displayed using the function:\n",
    "kipoi.print_dl_kwargs(DataLoader)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run dataloader on some examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fasta_file': 'example_files/hg38_chr22.fa',\n",
       " 'filter_protein_coding': True,\n",
       " 'gtf_file': 'example_files/gencode.v24.annotation_chr22.gtf',\n",
       " 'intervals_file': 'example_files/intervals.bed',\n",
       " 'target_file': 'example_files/targets.tsv'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# each dataloader already provides example files which can be used to illustrate its use:\n",
    "DataLoader.example_kwargs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd into the source directory \n",
    "os.chdir(DataLoader.source_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".\r\n",
      " custom_keras_objects.py -> ../template/custom_keras_objects.py\r\n",
      " dataloader_files\r\n",
      "  position_transformer.pkl\r\n",
      " dataloader.py -> ../template/dataloader.py\r\n",
      " dataloader.yaml -> ../template/dataloader.yaml\r\n",
      " example_files -> ../template/example_files\r\n",
      " model_files\r\n",
      "  model.h5\r\n",
      " model.yaml -> ../template/model.yaml\r\n",
      " __pycache__\r\n",
      "     custom_keras_objects.cpython-35.pyc\r\n",
      "     dataloader.cpython-35.pyc\r\n",
      "\r\n",
      "4 directories, 8 files\r\n"
     ]
    }
   ],
   "source": [
    "!tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "dl = DataLoader(**DataLoader.example_kwargs)\n",
    "# could be also done with DataLoader.init_example()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['metadata', 'inputs', 'targets'])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This particular dataloader is of type Dataset\n",
    "# i.e. it implements the __getitem__ method:\n",
    "dl[0].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['dist_gene_start', 'dist_gene_end', 'dist_stop_codon', 'dist_exon_intron', 'dist_start_codon', 'dist_tss', 'seq', 'dist_polya', 'dist_intron_exon'])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl[0][\"inputs\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.25, 0.25, 0.25, 0.25],\n",
       "       [0.25, 0.25, 0.25, 0.25],\n",
       "       [0.25, 0.25, 0.25, 0.25],\n",
       "       [0.25, 0.25, 0.25, 0.25],\n",
       "       [0.25, 0.25, 0.25, 0.25]], dtype=float32)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl[0][\"inputs\"][\"seq\"][:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the whole dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 1/1 [00:00<00:00,  3.29it/s]\n"
     ]
    }
   ],
   "source": [
    "whole_data = dl.load_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['metadata', 'inputs', 'targets'])"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_data.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 101, 4)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_data[\"inputs\"][\"seq\"].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the iterator to run predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "it = dl.batch_iter(batch_size=1, shuffle=False, num_workers=0, drop_last=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 101, 4)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(it)[\"inputs\"][\"seq\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.00050414]], dtype=float32)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict_on_batch(next(it)[\"inputs\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train the Keras model\n",
    "\n",
    "Keras model is stored under the `.model` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.model.compile(\"adam\", \"binary_crossentropy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_it = dl.batch_train_iter(batch_size=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/1\n",
      "3/3 [==============================] - 1s 270ms/step - loss: 1.0971\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.callbacks.History at 0x2ada28e3cba8>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.model.fit_generator(train_it, steps_per_epoch=3, epochs=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pipeline: `raw files -[dataloader]-> numpy arrays -[model]-> prediction`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_kwargs = model.default_dataloader.example_kwargs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [00:02,  2.67s/it]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1.        , 0.52299094, 0.52299094, 1.        , 1.        ,\n",
       "       1.        , 0.52299094, 1.        , 1.        , 1.        ,\n",
       "       1.        , 1.        , 1.        , 1.        ], dtype=float32)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.pipeline.predict(example_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.        ],\n",
       "       [0.52299076]], dtype=float32)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(model.pipeline.predict_generator(example_kwargs, batch_size=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:kipoi]",
   "language": "python",
   "name": "conda-env-kipoi-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
